src/components/abstractComponent.ts

```diff
@@ -0,0 +1,579 @@
+   * @returns {IComponent} The calling Component.
+   */
+  addClass(cssClass: string): this;
+  /**
+   * Removes a given CSS class from the Component.
+   *
+   * @param {string} cssClass The CSS class to remove.
+   * @returns {IComponent} The calling Component.
+   */
+  removeClass(cssClass: string): this;
+  /**
+   * Checks if the Component has a fixed width or if it grows to fill available space.
+   * Returns false by default on the base Component class.
+   */
+  fixedWidth(): boolean;
+  /**
+   * Checks if the Component has a fixed height or if it grows to fill available space.
+   * Returns false by default on the base Component class.
+   */
+  fixedHeight(): boolean;
+  /**
+   * Detaches a Component from the DOM. The Component can be reused.
+   *
+   * This should only be used if you plan on reusing the calling Component. Otherwise, use destroy().
+   *
+   * @returns The calling Component.
+   */
+  detach(): this;
+  /**
+   * Adds a callback to be called when the Component is detach()-ed.
+   *
+   * @param {GenericComponentCallback} callback
+   * @return {IComponent} The calling Component.
+   */
+  onDetach(callback: GenericComponentCallback<D>): this;
+  /**
+   * Removes a callback to be called when the Component is detach()-ed.
+   * The callback is identified by reference equality.
+   *
+   * @param {GenericComponentCallback} callback
+   * @return {IComponent} The calling Component.
+   */
+  offDetach(callback: GenericComponentCallback<D>): this;
+  /**
+   * Gets the parent ComponentContainer for this Component.
+   */
+  parent(): ComponentContainer;
+  /**
+   * Sets the parent ComponentContainer for this Component.
+   * An error will be thrown if the parent does not contain this Component.
+   * Adding a Component to a ComponentContainer should be done
+   * using the appropriate method on the ComponentContainer.
+   */
+  parent(parent: ComponentContainer): this;
+  parent(parent?: ComponentContainer): any;
+  /**
+   * @returns {Bounds} for the component in pixel space, where the topLeft
+   * represents the component's minimum x and y values and the bottomRight represents
+   * the component's maximum x and y values.
+   */
+  bounds(): Bounds;
+  /**
+   * Removes a Component from the DOM and disconnects all listeners.
+   */
+  destroy(): void;
+  /**
+   * Gets the width of the Component in pixels.
+   */
+  width(): number;
+  /**
+   * Gets the height of the Component in pixels.
+   */
+  height(): number;
+  /**
+   * Gets the origin of the Component relative to its parent.
+   *
+   * @return {Point}
+   */
+  origin(): Point;
+  /**
+   * Gets the origin of the Component relative to the root
+   *
+   * @return {Point}
+   */
+  originToRoot(): Point;
+  /**
+   * Gets the container holding the visual elements of the Component.
+   *
+   * Will return undefined if the Component has not been anchored.
+   *
+   * @return {IContent} content selection for the Component
+   */
+  content(): IContent;
```

Comment:
I didn't want to expose d3's api to end users. but I think its okay I'll just put that back